### ISSUE TITLE
createValidator, error with $type as array

### ISSUE THREAD

**Issue by XeuRun** [yiisoft/yii2#13269]
Method createValidator has check
```
if (is_array($type)) {
    $params = array_merge($type, $params);
} else {
    $params['class'] = $type;
}
```

### What steps will reproduce the problem?

```
public function rules()
{
    return [
        [
            'custom_field',
            ['class' => ValuesRemoteValidator::className()],
            'custom_url' => '',
            'custom_fields' => []
        ]
    ];
}
```

### What is the expected result?

Create validator with class ValuesRemoteValidator::className()

### What do you get instead?

Error, because `$model->hasMethod($type)` accepts a parameter $type as string, and `isset(static::$builtInValidators[$type]))` also cause an error

https://github.com/yiisoft/yii2/blob/2.0.10/framework/validators/Validator.php#L203

### Additional info

| Q                          | A
| ------------------- | ---
| Yii version             | 2.0.10
| PHP version          | 7.1.0
| Operating system | Ubuntu 16.04 x64

**Comment by arogachev** [comment]
You are attaching validator wrong. 

In case of configuring validator via array:

- 1st argument is always attribute name(s).
- 2nd parameter is the name / reference to actual validator.  For standalone validators implemented in separate classes, you need to pass class name as a string (not inside of array!). Example: `ValuesRemoteValidator::className()`.
- Starting from 3rd argument you can pass any custom parameters, such as `customUrl`, `customFields`, etc.

So correct option in your case will be:

```php
public function rules()
{
    return [
        [
            'custom_field',
            ValuesRemoteValidator::className(),
            'custom_url' => '',
            'custom_fields' => []
        ]
    ];
}
```

According function in `yii\base\Model`:

```php
public function createValidators()
{
    $validators = new ArrayObject;
    foreach ($this->rules() as $rule) {
        if ($rule instanceof Validator) {
            $validators->append($rule);
        } elseif (is_array($rule) && isset($rule[0], $rule[1])) { // attributes, validator type
            $validator = Validator::createValidator($rule[1], $this, (array) $rule[0], array_slice($rule, 2));
            $validators->append($validator);
        } else {
            throw new InvalidConfigException('Invalid validation rule: a rule must specify both attribute names and validator type.');
        }
    }
    return $validators;
}
```

As you can see from this line:

```php
if ($rule instanceof Validator) {
    $validators->append($rule);
} 
```

as an alternative to array you can pass the actual instance of used validator but it's not good idea to use it with rules because it will affect on readability and declaration style.

**Bottom line:** There is no error, it's expected behavior, you are attaching validator wrong. Issue can be closed.

**Comment by XeuRun** [comment]
@arogachev спасибо, я знаю как использовать кастомные валидаторы, но в коде
```
if (is_array($type)) {
    $params = array_merge($type, $params);
} else {
    $params['class'] = $type;
}
```
а точнее в проверке
`if (is_array($type))` переменная `$type` никогда не будет типа array, зачем нужна эта проверка?

**Comment by arogachev** [comment]
@XeuRun Sorry, but you wrote that you expect that validator will be created from that config.

**Comment by klimov-paul** [comment]
Validator type should nevere be an array - it make absolutely no sense as all validation params are already a validator configuration:

```
public function rules()
{
    return [
        [
            'custom_field',
            ValuesRemoteValidator::className(),
            'custom_url' => '', // match filed/property at `ValuesRemoteValidator`
            'custom_fields' => [] // match filed/property at `ValuesRemoteValidator`
        ]
    ];
}
```

**Comment by arogachev** [comment]
@klimov-paul Actually he means that this check:

```php
if (is_array($type)) {
    $params = array_merge($type, $params);
}
```

in the `createValidator()` method of a `yii\validators\Validator` class is redundant exactly because of the reasons we both described above.

Initial formulation is a bit confusing, but now I see what he is talking about.

Indeed, it seems to be redundant. And by the way it's not documented in PHPDoc:

```php
* @param string|\Closure $type the validator type. This can be either:
*  * a built-in validator name listed in [[builtInValidators]];
*  * a method name of the model class;
*  * an anonymous function;
*  * a validator class name.
```

Allowed types are string and closure for now.

**Comment by klimov-paul** [comment]
`Validator::createValidator()` can be used standalone without model rules involved. That is why it contains such a check.
However `Model::rules()` result is processed in its own way, which is explicitely documented.

**Comment by arogachev** [comment]
Yes, but we have a 4th argument called `params` and can pass any additional params through it. So we have 2 ways of initialization:

```php
Validator::createValidator(['class' => SomeValidator::className(), 'foo' => 'bar'], $model, 'attribute1', []);
```

and 

```php
Validator::createValidator(SomeValidator::className(), $model, 'attribute1', ['foo' => 'bar']);
```

Is it OK?

**Comment by klimov-paul** [comment]
Yes

**Comment by XeuRun** [comment]
@arogachev @klimov-paul 

initialization by first way: 
```
Validator::createValidator(['class' => SomeValidator::className(), 'foo' => 'bar'], $model, 'attribute1', []);
```
it causes an error in the
`if ($type instanceof \Closure || $model->hasMethod($type))`
because
`$model->hasMethod($type)`
accept
`* @param string $name the property name`
